Use dynamicDowncast<T> more in css/

cdumez opened the pull request:
#### 1540d326ead095e4efd8cf6f46fe6068f84acacd
<pre>
Use dynamicDowncast&lt;T&gt; more in css/
<a href="https://bugs.webkit.org/show_bug.cgi?id=265219">https://bugs.webkit.org/show_bug.cgi?id=265219</a>

Reviewed by Jean-Yves Avenard.

* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::buildSerializablePositionOffset):
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::symbolFromCSSValue):
(WebCore::nameFromCSSValue):
* Source/WebCore/css/CSSCursorImageValue.cpp:
(WebCore::CSSCursorImageValue::create):
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::calculateWeightRange):
(WebCore::calculateStretchRange):
(WebCore::calculateItalicRange):
(WebCore::CSSFontFace::setFeatureSettings):
(WebCore::CSSFontFace::document):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
(WebCore::CSSFontFaceSource::isSVGFontFaceSource const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFaceRule):
(WebCore::CSSFontSelector::updateStyleIfNeeded):
(WebCore::CSSFontSelector::fallbackFontAt):
* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::mutableProperties):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::valueID):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::generateSelectorText const):
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::nestedRules const):
(WebCore::CSSStyleRule::reattach):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::modifiedKeyframesRuleName):
(WebCore::CSSStyleSheet::RuleMutationScope::RuleMutationScope):
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillRepeat):
(WebCore::CSSToStyleMap::mapAnimationDelay):
(WebCore::CSSToStyleMap::mapAnimationDuration):
(WebCore::CSSToStyleMap::mapAnimationIterationCount):
(WebCore::CSSToStyleMap::mapAnimationName):
(WebCore::CSSToStyleMap::mapAnimationProperty):
(WebCore::CSSToStyleMap::mapNinePieceImage):
(WebCore::CSSToStyleMap::mapNinePieceImageSlice):
(WebCore::CSSToStyleMap::mapNinePieceImageWidth):
(WebCore::CSSToStyleMap::mapNinePieceImageQuad):

Canonical link: <a href="https://commits.webkit.org/271065@main">https://commits.webkit.org/271065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18f6c5ae279233059a30a28a05f3118977020354

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27214 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5853 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28462 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29437 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24896 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27682 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24732 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27476 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4643 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23353 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4055 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4162 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30076 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24844 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30345 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2349 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28273 "Passed tests") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5665 "Hash 18f6c5ae for PR 20804 does not build (failure)") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6559 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4658 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4579 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -91,9 +91,9 @@ static SerializablePositionOffset buildSerializablePositionOffset(CSSValue* offs
     if (side == CSSValueCenter)
         side = defaultSide;
-    else if ((side == CSSValueRight || side == CSSValueBottom) && is<CSSPrimitiveValue>(*amount) && downcast<CSSPrimitiveValue>(*amount).isPercentage()) {
+    else if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(*amount); (side == CSSValueRight || side == CSSValueBottom) && primitiveValue && primitiveValue->isPercentage()) {
```
this allows to keep a one-line, but it changes the order of operation, do we care about performance for things like this here?

jyavenard commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -571,12 +571,18 @@ CSSStyleSheet::RuleMutationScope::RuleMutationScope(CSSStyleSheet* sheet, RuleMu
     , m_contentsWereClonedForMutation(ContentsWereNotClonedForMutation)
     , m_insertedKeyframesRule(nullptr)
-    , m_modifiedKeyframesRuleName(is<CSSKeyframesRule>(rule) ? downcast<CSSKeyframesRule>(*rule).name() : emptyAtom())
+    , m_modifiedKeyframesRuleName(modifiedKeyframesRuleName(rule))
```
I think a lambda is more elegant here.

cdumez replied:
A multi line lambda in the middle of the members initialization? Not convinced this will look better than my static function?

cdumez replied:
I made the change, not sure if it's better but why not.

jyavenard replied:
It lets you see what’s going on at a glance. Without having to look for the method. Though here, it’s right next to it.

cdumez committed:
Use dynamicDowncast<T> more in css/
https://bugs.webkit.org/show_bug.cgi?id=265219

Reviewed by Jean-Yves Avenard.

* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::buildSerializablePositionOffset):
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::symbolFromCSSValue):
(WebCore::nameFromCSSValue):
* Source/WebCore/css/CSSCursorImageValue.cpp:
(WebCore::CSSCursorImageValue::create):
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::calculateWeightRange):
(WebCore::calculateStretchRange):
(WebCore::calculateItalicRange):
(WebCore::CSSFontFace::setFeatureSettings):
(WebCore::CSSFontFace::document):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
(WebCore::CSSFontFaceSource::isSVGFontFaceSource const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFaceRule):
(WebCore::CSSFontSelector::updateStyleIfNeeded):
(WebCore::CSSFontSelector::fallbackFontAt):
* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::mutableProperties):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::valueID):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::generateSelectorText const):
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::nestedRules const):
(WebCore::CSSStyleRule::reattach):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::modifiedKeyframesRuleName):
(WebCore::CSSStyleSheet::RuleMutationScope::RuleMutationScope):
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillRepeat):
(WebCore::CSSToStyleMap::mapAnimationDelay):
(WebCore::CSSToStyleMap::mapAnimationDuration):
(WebCore::CSSToStyleMap::mapAnimationIterationCount):
(WebCore::CSSToStyleMap::mapAnimationName):
(WebCore::CSSToStyleMap::mapAnimationProperty):
(WebCore::CSSToStyleMap::mapNinePieceImage):
(WebCore::CSSToStyleMap::mapNinePieceImageSlice):
(WebCore::CSSToStyleMap::mapNinePieceImageWidth):
(WebCore::CSSToStyleMap::mapNinePieceImageQuad):

Canonical link: https://commits.webkit.org/271065@main
